src/sqlfluff/dialects/dialect_ansi.py

```diff
@@ -1460,7 +1460,10 @@ def get_eventual_aliases(self) -> List[Tuple[BaseSegment, AliasInfo]]:
             buff.append((from_expression, alias))
 
         # In some dialects, like TSQL, join clauses can have nested join clauses
-        for join_clause in self.recursive_crawl("join_clause"):
+        # recurse into them - but not if part of a sub-select statement (see #3144)
+        for join_clause in self.recursive_crawl(
```

Comment:
I don't think we're fixing the real issue here. It seems to me that we don't need to ignore `select_statement`. Instead, we need to detect that the `on` clause is using the alias, therefore it's not unused.

Do we know why the behavior changed from previous versions?

Reply:
This was discussed more in [this comment in #3144](https://github.com/sqlfluff/sqlfluff/issues/3144#issuecomment-1105721052)

Basically, previously it was this:

```py
for join_clause in self.get_children("join_clause"):
```

That is, we only got the direct join children, rather than using `recursive_crawl`.

However #2993 was added to make it recursive to allow SQL like this:

```sql
    SELECT
        1 AS RegionCode
    FROM BA
    LEFT OUTER JOIN (
        I LEFT OUTER JOIN P
            ON I.Pcd = P.Iid
    ) ON BA.Iid = I.Bcd
```

Note that the outermost `ON` clause is referencing the inner table `I`. This is a specific syntax allowed in T-SQL (but no where else I don't think?), so the above code needed to change to be `recursive_crawl` instead of `get_children`.

Unfortunately there is no easier way of identifying just this case as there will be brackets, and there could be nested, nested calls. But `recursive_crawl` will find them so all good.

That solve that problem, but then broke it for this SQL - hence this PR to partially revert that:

```sql
select col1
from table1 as abg
inner join (
    select
        ceb.emailaddresskey,
        dac.accountkey
    from table2 as dac
    inner join table3 as ceb
        on ceb.col2 = dac.col2
) as abh
on abg.col1 = abh.col1
```

Here the subquery should **not** be accessible at all to the outer query, and only the results of that (as `abh`) as it's not that specific T-SQL instance above, which is the only one that should allow nesting. So effectively it should be a black box, and should look like this (at least while while the rule process the outer select - it'll get to the nested one later):

```sql
select col1
from table1 as abg
inner join (
    ---something we don't need to worry about now
) as abh
on abg.col1 = abh.col1
```

Then we process that inner join later in the rule.

That is what used to happen before #2993 , and then, while processing the outer select statement,  it would see table references `abg` and `abh` and then uses of both those in the `ON` clause - great, no rule to fire!

Instead what happens (and why we need this PR) is we now pick up recursive "joins" so we ALSO pick up table reference `ceb` when we shouldn't. And there is no reference to `ceb` in the outer select, hence the rule fires complaining about that.

Note that we don't pick up the inner table reference `dac` btw cause that is an ``FROM` clause and not a `JOIN` clause and the bug is specific to `JOIN` clauses. Again, showing that this is wrong to traverse into the inner select for just `JOIN`s.

As the nested `SELECT` should be a black box (except in very specific use cases for T-SQL, when it's an nested `JOIN`), I've added functionality to **not** recurse into it anymore, if it is it's own select statement (which is exactly what happened before), but still recurse for anything else (so the original T-SQL bug is still resolved). There may be other instances we might need to do that, but that's the only one I can think of now.

The alternative is to gather ALL the table references for the whole query (even the inner, nested ones that should not be referenceable, and including `dac` in this example) and ALL the column references, and then decide which are used and which are not. But that is not the correct thing to do here, since the inner, nested table references are not be accessible outside of it.

Here's a BigQuery example showing the inner table canot be referenced in the outer select:

<img width="609" alt="image" src="https://user-images.githubusercontent.com/10931297/164621142-9ef1d322-581a-4e7a-b957-21abb3d9c54e.png">

Only via the outer alias:

<img width="610" alt="image" src="https://user-images.githubusercontent.com/10931297/164621420-b4c74d94-94f7-45a6-aa3b-45d79f83b06a.png">

Does that explain it all?

Reply:
This makes sense to me, but I'll leave it to @barrywhart to ✅ if he is happy with the explanation

Reply:
Makes sense!

In general, visibility goes upwards, but not downwards.